### PR TITLE
Ds 675 hat to adj auth

### PR DESCRIPTION
**Pull request by dataswifty**

### CODE REVIEW

File: hat/app/org/hatdex/hat/api/controllers/RichData.scala [hunk 1]
```diff
@@ -83,8 +83,11 @@ class RichData @Inject() (
     configuration.underlying.getString("adjudicator.address")
   val adjudicatorScheme =
     configuration.underlying.getString("adjudicator.scheme")
-  val adjudicatorEndpoint = s"${adjudicatorScheme}${adjudicatorAddress}"
-  val adjudicatorClient = new AdjudicatorRequest(adjudicatorEndpoint, wsClient)
+  val adjudicatorEndpoint =
```
Comment (benshaw) [inline hat/app/org/hatdex/hat/api/controllers/RichData.scala:86]: what is the appropriate access for these ? do you really want them public ? If so public members should have an explicit type https://dswift.atlassian.net/wiki/spaces/DSENGINEER/pages/4751401/Best+Practices#Use-explicit-types

File: hat/app/org/hatdex/hat/utils/AdjudicatorRequest.scala [hunk 2]
```diff
@@ -50,49 +56,73 @@ object AdjudicatorRequestTypes {
...
 
-  def getPublicKey(hatName: String, contractId: ContractId, keyId: String)(implicit ec: ExecutionContext): Future[Either[PublicKeyRequestFailure, PublicKeyReceived]] = {
+  //
+  private val secret = JwtSecretKey(sharedSecret)
+  val hatClaimBuiler: JwtClaimBuilder[HatClaim] = HatClaim.builder
 
-    val url =
-      s"${adjudicatorEndpoint}/v1/contracts/${contractId}/hat/${hatName}/${keyId}"
-    val eventuallyResponse = makeRequest(url, ws).get()
+  // Create HatClaim
```
Comment (benshaw) [inline hat/app/org/hatdex/hat/utils/AdjudicatorRequest.scala:66]: redundant comment

File: hat/app/org/hatdex/hat/utils/AdjudicatorRequest.scala [hunk 3]
```diff
@@ -50,49 +56,73 @@ object AdjudicatorRequestTypes {
...
+  private def createHatClaim(hatNameAsStr: String, contractId: ContractId): Option[HatClaim] = {
```
Comment (benshaw) [inline hat/app/org/hatdex/hat/utils/AdjudicatorRequest.scala:67]: Avoid method overloading (https://dswift.atlassian.net/wiki/spaces/DSENGINEER/pages/4751401/Best+Practices#Avoid-Method/Function-overloading)

**Comment by dataswifty** [comment]
I need to mock the tests.

File: hat/app/org/hatdex/hat/utils/AdjudicatorRequest.scala [hunk 4]
```diff
@@ -50,49 +56,73 @@ object AdjudicatorRequestTypes {
...
-                "The response was not able to be decoded into an Array[Byte]."))
-            case Some(arrayByte) => Right(PublicKeyReceived(arrayByte))
-          }
-        }
-        case _ => {
-          Left(PublicKeyRequestFailure.ServiceRespondedWithFailure(
-            s"The Adjudicator Service responded with an error: ${response.statusText}"))
-        }
+  private def createHatClaim(maybeHatName: Option[HatName], contractId: ContractId): Option[HatClaim] = {
+    for {
```
Comment (benshaw) [inline hat/app/org/hatdex/hat/utils/AdjudicatorRequest.scala:75]: wondering if a for comp is best here
would  ``` maybeHatName.map(HatClaim(_, contractId)) ``` read better ?
is this repeated enough that it really need a function ?

File: hat/app/org/hatdex/hat/utils/AdjudicatorRequest.scala [hunk 5]
```diff
@@ -50,49 +56,73 @@ object AdjudicatorRequestTypes {
...
+      hatName <- maybeHatName
+      hatClaim = HatClaim(hatName, contractId)
+    } yield hatClaim
+  }
+
+  // Internal calls
+  def getPublicKey(hatName: HatName, contractId: ContractId, keyId: String)(implicit ec: ExecutionContext): Future[Either[PublicKeyRequestFailure, PublicKeyReceived]] = {
+    val claim = createHatClaim(Some(hatName), contractId)
```
Comment (benshaw) [inline hat/app/org/hatdex/hat/utils/AdjudicatorRequest.scala:83]: Is a function needed ? is ```Some(HatClaim(hatName,contractId))```  clearer

File: hat/app/org/hatdex/hat/utils/AdjudicatorRequest.scala [hunk 6]
```diff
@@ -50,49 +56,73 @@ object AdjudicatorRequestTypes {
...
 }
 
-class AdjudicatorRequest(adjudicatorEndpoint: String, ws: WSClient) {
+class AdjudicatorRequest(adjudicatorEndpoint: String, sharedSecret: String, ws: WSClient) {
   import AdjudicatorRequestTypes._
 
-  def getPublicKey(hatName: String, contractId: ContractId, keyId: String)(implicit ec: ExecutionContext): Future[Either[PublicKeyRequestFailure, PublicKeyReceived]] = {
+  //
+  private val secret = JwtSecretKey(sharedSecret)
+  val hatClaimBuiler: JwtClaimBuilder[HatClaim] = HatClaim.builder
```
Comment (benshaw) [inline hat/app/org/hatdex/hat/utils/AdjudicatorRequest.scala:64]: should this be private

File: hat/app/org/hatdex/hat/utils/AdjudicatorRequest.scala [hunk 7]
```diff
@@ -50,49 +56,73 @@ object AdjudicatorRequestTypes {
   final case class ContractLeft(contractId: ContractId)
 }
 
-class AdjudicatorRequest(adjudicatorEndpoint: String, ws: WSClient) {
+class AdjudicatorRequest(adjudicatorEndpoint: String, sharedSecret: String, ws: WSClient) {
```
Comment (benshaw) [inline hat/app/org/hatdex/hat/utils/AdjudicatorRequest.scala:59]: could sharedSecret be passes as type JwtSecretKey ?

File: hat/app/org/hatdex/hat/utils/AdjudicatorRequest.scala [hunk 8]
```diff
@@ -50,49 +56,73 @@ object AdjudicatorRequestTypes {
...
+    val claim = createHatClaim(hatName, contractId)
+
+    val url = s"${adjudicatorEndpoint}/v1/contracts/hat"
+
+    makeRequest(url, claim, ws) match {
+      case Some(req) => {
+        runJoinContractRequest(req, contractId)
       }
+      case None => Future.successful(Left(JoinContractRequestFailure.ServiceRespondedWithFailure(
+        s"The Adjudicator Service responded with an error.")))
```
Comment (benshaw) [inline hat/app/org/hatdex/hat/utils/AdjudicatorRequest.scala:106]: DRY this pattern inc error is repeated everywhere could it potentially be extracted ?

**Comment by dataswifty** [comment]
This is also depends on releases of adj-https and adj-lib

**Pull request merged**